### DISCUSSION TITLE
Installation on Windows

### DISCUSSION THREAD

**Discussion by jfbribeiro** [Yubico/libfido2#845] [General]
I need a clear explanation on how to install libfido2 on my Windows 64 machine.
I unzipped the file which contains 5 folders ( ARM, ARM64, include, Win32 and Win64), but now I'm lost.

**Comment by LDVG** [comment]
Hi,

To answer your question, please elaborate on your use case.  Are you building an application that interfaces with libfido2 (which would be the main use case)?

**Reply by jfbribeiro** [reply]
Yes, I am building an app that interfaces with libfido2.
The idea is that the app looks for a connected NFC reader, asks for a FIDO2 authentication and sends the received data to the relying party.
So far I'm doing it like this (which I don't know if is the right way so far), but I would like to install libfido2 on my windows machine so that I don't have to use the fido2-token and fido2-assert executables.
```
var token = 'pathTo//fido2-token.exe';
  var assertX = ''pathTo//fido2-assert.exe';

  var args = ['-L'];

   ProcessResult result = await runExecutableArguments(token, args);

   String deviceId = (result.stdout as String).split(" ")[0];

   args = ['-G', '-r', 'outlook.com', device];
   result = await runExecutableArguments(assertX, args);
```

**Reply by LDVG** [reply]
OK. Typically, you would instruct your build system to link with libfido2 in your preferred manner. You may want to vendor our DLL(s) when you distribute your application, or instruct your users to make our DLLs available in their path, or statically link. I'm afraid there's not really any "one-size-fits-all" install instruction.

Given the use of `runExecutableArguments()`, I'm assuming the above is Dart? I'm not familiar with it myself, but if I were you I'd start by looking at [their FFI documentation](https://dart.dev/interop/c-interop#bundle-and-load-c-libraries).

**Reply by jfbribeiro** [reply]
Thanks for the explanation. So far I made the DLLs available in my path and I'm using Dart FFI to replicate the manifest that you have in your examples and try to look for authenticator devices. 
I have an Yubikey connected, but it seems that the only thing that is found is my own Windows machine,


```
import 'dart:ffi'; // FFI
import 'package:ffi/ffi.dart'; // For memory management

// Load the dynamic library
final dylib = DynamicLibrary.open('windows\\lib\\fido2.dll'); // or .so/.dylib

// FFI bindings for FIDO2 functions and structs
typedef FidoInitNative = Void Function(Uint32 flags);
typedef FidoInitDart = void Function(int flags);

// Bind fido_init
final fidoInit = dylib.lookupFunction<FidoInitNative, FidoInitDart>('fido_init');

// Bind fido_dev_info_new
typedef FidoDevInfoNewNative = Pointer<FidoDevInfo> Function(Uint64 n);
typedef FidoDevInfoNewDart = Pointer<FidoDevInfo> Function(int n);
final fidoDevInfoNew = dylib.lookupFunction<FidoDevInfoNewNative, FidoDevInfoNewDart>('fido_dev_info_new');

// Bind fido_dev_info_manifest
typedef FidoDevInfoManifestNative = Int32 Function(Pointer<FidoDevInfo>, Uint64, Pointer<Uint64>);
typedef FidoDevInfoManifestDart = int Function(Pointer<FidoDevInfo>, int, Pointer<Uint64>);
final fidoDevInfoManifest = dylib.lookupFunction<FidoDevInfoManifestNative, FidoDevInfoManifestDart>('fido_dev_info_manifest');

// Additional bindings: fido_dev_info_ptr, vendor, product, etc.
typedef FidoDevInfoPtrNative = Pointer<FidoDevInfo> Function(Pointer<FidoDevInfo>, Uint64);
typedef FidoDevInfoPtrDart = Pointer<FidoDevInfo> Function(Pointer<FidoDevInfo>, int);
final fidoDevInfoPtr = dylib.lookupFunction<FidoDevInfoPtrNative, FidoDevInfoPtrDart>('fido_dev_info_ptr');

// Dart representation of the fido_dev_info_t struct
base class FidoDevInfo extends Struct {
  external Pointer<Utf8> path;
  @Int16()
  external int vendor_id;
  @Int16()
  external int product_id;
  external Pointer<Utf8> manufacturer;
  external Pointer<Utf8> product;
}

void main() {
  fidoInit(0);

  int deviceCount = 64;
  Pointer<FidoDevInfo> devList = fidoDevInfoNew(deviceCount);

  if (devList.address == 0) {
    print('Failed to allocate device list.');
    return;
  }

  final nDevsPtr = calloc<Uint64>(); // To store the number of detected devices
  int result = fidoDevInfoManifest(devList, deviceCount, nDevsPtr);
  if (result != 0) { // Assuming FIDO_OK is 0
    print('Device manifest error: $result');
    calloc.free(nDevsPtr);
    return;
  }

  int nDevs = nDevsPtr.value;
  print('Number of devices detected: $nDevs');

  for (int i = 0; i < nDevs; i++) {
    final devicePtr = fidoDevInfoPtr(devList, i).ref;

    String path = devicePtr.path.toDartString();
    String manufacturer = devicePtr.manufacturer != nullptr ? devicePtr.manufacturer.toDartString() : 'Unknown';
    String product = devicePtr.product != nullptr ? devicePtr.product.toDartString() : 'Unknown';

    print('Device $i:');
    print('  Path: $path');
    print('  Vendor ID: ${devicePtr.vendor_id}');
    print('  Product ID: ${devicePtr.product_id}');
    print('  Manufacturer: $manufacturer');
    print('  Product: $product');
  }

  calloc.free(nDevsPtr);
}

```

And the output is : 

```
Number of devices detected: 1
Device 0:
  Path: windows://hello
  Vendor ID: 1118
  Product ID: 1
  Manufacturer: Microsoft Corporation
  Product: Windows Hello
```

Am I missing something here ? Some configuration ?

**Reply by LDVG** [reply]
Happy to hear that it helped.

> Am I missing something here ? Some configuration ?

The Windows Hello pseudo-device functions as a proxy for HID communication, which is blocked to unprivileged applications in Windows. Using the pseudo-device, you can create new credentials and get assertions from any authenticator connected to the computer.  If you want to speak directly to the authenticator, you will have to run your application with escalated (administrator) privileges.

**Reply by jfbribeiro** [reply]
So in my case, and considering that I have an NFC card reader connected through USB and I'm able to find the pseudo-device and ask him for assertions, I can expect that when I ask for that assertion it can come from the card reader, right ?

**Reply by LDVG** [reply]
I'm afraid I'm unable to answer questions about specific readers -- it's essentially up to the implementation of webauthn.dll whether or not it will support it. :slightly_smiling_face:
